### PR TITLE
feat(sdk): state transitions support in rust sdk

### PR DESCRIPTION
**Pull request by lklimek**
## Issue being fixed or feature implemented
This PR is for state transition support in dash-sdk. This means the ability for a client to put documents through state transitions to platform and have them be stored on the blockchain. As well as a lot of additional query support.

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [ ] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

File: packages/rs-sdk/src/core_client.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            .get_balance(None, None)
+            .map_err(Error::CoreClientError)
+    }
+
+    /// Retrieve quorum public key from core.
+    pub fn get_quorum_public_key(
+        &self,
+        quorum_type: u32,
+        quorum_hash: [u8; 32],
+        _core_chain_locked_height: u32,
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-sdk/src/core_client.rs:109]: I mean _core_chain_locked_height: u32, if it isn't being used.

File: packages/rs-drive-proof-verifier/src/provider.rs [hunk 2]
```diff
@@ -54,36 +122,32 @@ impl MockQuorumInfoProvider {
...
 }
 
 #[cfg(feature = "mocks")]
-impl QuorumInfoProvider for MockQuorumInfoProvider {
-    /// Mock implementation of [QuorumInfoProvider] that returns keys from files saved on disk.
+impl ContextProvider for MockContextProvider {
+    /// Mock implementation of [ContextProvider] that returns keys from files saved on disk.
     ///
-    /// Use `rs_sdk::SdkBuilder::with_dump_dir()` to generate quorum keys files.   
+    /// Use `dash_platform_sdk::SdkBuilder::with_dump_dir()` to generate quorum keys files.
```
Comment (lklimek) [inline packages/rs-drive-proof-verifier/src/provider.rs:141]: is is rs-sdk or dash_platform_sdk? please check all renames

File: packages/rs-sdk/src/core_client.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-sdk/src/core_client.rs:109]: agree

File: packages/rs-sdk/src/platform/put.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
+//! Module containing the Put trait, allowing modification of objects (state transitions) on the Dash Platform.
```
Comment (lklimek) [inline packages/rs-sdk/src/platform/put.rs:1]: this file is unused as state transitions turned out to be much more complex than I expected. To be removed.

**Review by lklimek — COMMENTED** [review]
LGTM, some comments added that can be handled in separate PR

File: packages/rs-drive-abci/src/execution/platform_events/identity_credit_withdrawal/fetch_and_prepare_unsigned_withdrawal_transactions/v0/mod.rs [hunk 4]
```diff
@@ -0,0 +1,169 @@
+use dashcore_rpc::dashcore::{
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-drive-abci/src/execution/platform_events/identity_credit_withdrawal/fetch_and_prepare_unsigned_withdrawal_transactions/v0/mod.rs:1]: delete this file, looks like leftover

File: packages/rs-sdk/Cargo.toml [hunk 5]
```diff
@@ -8,19 +8,22 @@ dpp = { path = "../rs-dpp", features = [
...
 ] }
 dapi-grpc = { path = "../dapi-grpc", features = ["client"] }
 rs-dapi-client = { path = "../rs-dapi-client", default-features = false }
 drive = { path = "../rs-drive", default-features = false, features = [
     "verify",
 ] }
 drive-proof-verifier = { path = "../rs-drive-proof-verifier" }
+dapi-grpc-macros = { path = "../rs-dapi-grpc-macros" }
 # drive-abci is only needed for core rpc; TODO remove once we have correct core rpc impl
 drive-abci = { path = "../rs-drive-abci", default-features = false }
```
Comment (lklimek) [inline packages/rs-sdk/Cargo.toml:21]: please address `cargo machete` output, https://github.com/dashpay/platform/actions/runs/7828491935/job/21358511981?pr=1596 (needs to be double-checked as machete is not perfect, for example when sth is hidden behind feature)

File: packages/rs-drive-abci/Cargo.toml [hunk 6]
```diff
@@ -35,15 +33,16 @@ bs58 = "0.4.0"
 hex = "0.4.3"
 indexmap = { version = "1.9.3", features = ["serde"] }
 sha2 = "0.10.6"
-dashcore-rpc = { git = "https://github.com/dashpay/rust-dashcore-rpc", tag = "v0.15.2" }
+dashcore-rpc = { git = "https://github.com/dashpay/rust-dashcore-rpc", branch = "master" }
```
[the remark this answers is not in the document]
Comment (fominok) [reply packages/rs-drive-abci/Cargo.toml:36]: good catch, no idea how it comes back every time like that

**Comment by QuantumExplorer** [comment]
Awesome work from all involved, very happy to see this one merged in.

**Pull request merged**